### DISCUSSION TITLE
Feedback on Express HTTP + Websockets server

### DISCUSSION THREAD

**Discussion by tkroll** [trpc/trpc#1980] [Q&A] [answered]
I put together this Express http/ws server. It seems to work fine.
Any improvements or comments? 
**Thanks**

```typescript
import cors from 'cors'
import express from 'express'
import http from 'http'
import { Server } from 'ws'

import { AppRouter, appRouter as router, createContext } from '@lib/trpc'
import { createExpressMiddleware } from '@trpc/server/adapters/express'
import { applyWSSHandler } from '@trpc/server/adapters/ws'

// http server
const app = express()
const server = http.createServer(app)

// web socket server
const wss = new Server({ server })
const wsHandler = applyWSSHandler<AppRouter>({
  wss,
  router,
  createContext,
})

app.use(cors())
app.use(
  '/trpc',
  createExpressMiddleware<AppRouter>({
    router,
    createContext,
  }),
)

app.get('/api', (req, res) => {
  res.send({ message: 'Welcome to trpc!' })
})

const port = process.env.TRPC_PORT || 8123
server.listen(port, () => {
  console.log(`Listening at http://localhost:${port}`)
})
server.on('error', console.error)

process.on('SIGTERM', () => {
  wsHandler.broadcastReconnectNotification()
  wss.close()
  server.close()
})
```

**Comment by jlalmes** [comment, marked as answer]
Looks good to me! ✅

**Reply by tkroll** [reply]
Thanks for taking a look. My Express is a bit rusty.

**Reply by tkroll** [reply]
Late follow-up to my original.

Notice for this Express server I'm using `CreateNextContextOptions`. I overlooked this when slapping the code together.
Recently tried to replace it with the more correct `CreateExpressContextOptions`, but this makes the `createContext` in the server's `applyWSSHandler ` complain.

What is the correct way to get this working without using Next stuff in an Express server?

Thanks!

```ts
import { inferAsyncReturnType } from '@trpc/server'
import { CreateNextContextOptions } from '@trpc/server/adapters/next'
// import { CreateExpressContextOptions } from '@trpc/server/adapters/express'

import { prisma } from '../../prisma/prisma'

export const createContext = ({ req, res }: CreateNextContextOptions) => {
  return {
    req,
    res,
    prisma
  }
}

export type Context = inferAsyncReturnType<typeof createContext>
```

**Reply by KATT** [reply]
Try 

```ts
export const createContext = ({ req, res }: CreateNextContextOptions | CreateExpressContextOptions) => {
```

**Reply by brian-chano** [reply]
> ```ts
> import { CreateNextContextOptions } from '@trpc/server/adapters/next'
> ```

Hej @KATT ,

Any idea on how to type context when I am using websockets and express-session?

I am adding a userId to my session and typing this as 
```typescript

declare module 'http' {
  interface IncomingMessage {
    session: Session & {
      userId?: string
    }
  }
}

export const createContext = async ({req,res}: CreateExpressContextOptions | CreateWSSContextFnOptions) => {}
```

but I get undefined when I call req.session.userId in my procedures

**Reply by mocon** [reply]
Same issue here with Express context causing the TS error with createContext.

**Reply by ferm10n** [reply]
When creating a websocket context for trpc, the `opts.req` doesn't seem to be coming from express, so it makes sense that `opts.req.session` is undefined.

My solution for this was to read the cookie in the createContext for the websocket, and pull data from a store using the id in the cookie.
